zend_abstract_interface/properties/tests/php7-8/properties.cc

```diff
@@ -0,0 +1,134 @@
+
+    zval_ptr_dtor(&super);
+})
+
+TEST("access public property", {
+    zval super;
+    zai_call_function_literal("zai\\properties\\test\\super", &super);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(super), Z_OBJ(super), "public");
+    REQUIRE(Z_TYPE_P(val) == IS_STRING);
+    REQUIRE(zend_string_equals_literal(Z_STR_P(val), "public"));
+
+    zval_ptr_dtor(&super);
+})
+
+TEST("access overridden private property on child", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child), Z_OBJ(child), "private");
+    REQUIRE(Z_TYPE_P(val) == IS_STRING);
+    REQUIRE(zend_string_equals_literal(Z_STR_P(val), "private from child"));
+
+    zval_ptr_dtor(&child);
+})
+
+TEST("access overridden private property on parent", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child)->parent, Z_OBJ(child), "private");
+    REQUIRE(Z_TYPE_P(val) == IS_STRING);
+    REQUIRE(zend_string_equals_literal(Z_STR_P(val), "private"));
+
+    zval_ptr_dtor(&child);
+})
+
+TEST("access private property on parent", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child)->parent, Z_OBJ(child), "superPrivate");
+    REQUIRE(Z_TYPE_P(val) == IS_STRING);
+    REQUIRE(zend_string_equals_literal(Z_STR_P(val), "superPrivate"));
+
+    zval_ptr_dtor(&child);
+})
+
+TEST("access public property of parent", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child), Z_OBJ(child), "public");
+    REQUIRE(Z_TYPE_P(val) == IS_STRING);
+    REQUIRE(zend_string_equals_literal(Z_STR_P(val), "public"));
+
+    zval_ptr_dtor(&child);
+})
+
+TEST("access dynamic property", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child), Z_OBJ(child), "dynamicPrivate");
+    REQUIRE(Z_TYPE_P(val) == IS_STRING);
+    REQUIRE(zend_string_equals_literal(Z_STR_P(val), "dynamicPrivate from child"));
+
+    zval_ptr_dtor(&child);
+})
+
+TEST("access inexistent property is null", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child), Z_OBJ(child), "undefined");
+    REQUIRE(val == &EG(uninitialized_zval));
+
+    zval_ptr_dtor(&child);
+})
+
+TEST("access property with invalid name is error", {
+    zval child;
+    zai_call_function_literal("zai\\properties\\test\\child", &child);
+
+    zval *val = zai_read_property_direct_literal(Z_OBJCE(child), Z_OBJ(child), "\0with zero byte");
+    REQUIRE(val == &EG(error_zval));
+
+    zval_ptr_dtor(&child);
+})
```

Comment:
In the spirit of being super defensive, could we add some tests where `ce`, `object`, and `name` args are `NULL`?

Also, do you think it's worth it to add some tests to access properties on instances of internal classes as well?

Reply:
I agree with ensuring the caller does this check in principle, but until it can be enforced by static analysis or something, I would prefer to be defensive. For you and I getting a sigsegv directly at the site is nice, but for customers and their customers it's not good at all, as crashes bring down the process.

Reply:
Actually, we probably can just log that instead? I just have no idea what the proper logging within ZAI would be for that? ZAI has no access to the standard ddtrace_log functions

Reply:
Re internal class access: there is no difference in the zend_object * structure whether something is internal or not. It just might be that internal classes have some data attached and may define their own read_properties, but this function does not support that use case anyway.